### ISSUE TITLE
OCI Distribution Conformance Testing

### ISSUE THREAD

**Issue by upodroid** [kubernetes/registry.k8s.io#129]
The OCI Distribution spec has a conformance test that you can run to see if a registry meets the OCI spec.

I ran this for archeio and we are doing great with blobs but not so well with the manifests.

https://github.com/opencontainers/distribution-spec/tree/main/conformance

Set the following variables:

```
export OCI_NAMESPACE=kubernetes/pause
export OCI_TAG_NAME=3.7
export OCI_TEST_PULL=1
export OCI_ROOT_URL=https://registry.k8s.io
export OCI_TEST_PUSH=0
export OCI_MANIFEST_DIGEST=sha256:f81611a21cf91214c1ea751c5b525931a0e2ebabe62b3937b6158039ff6f922d
export OCI_BLOB_DIGEST=sha256:221177c6082a88ea4f6240ab2450d540955ac6f4d5454f0e15751b653ebda165
```

We need to change a few things to make it comply with the spec.

Report link: https://htmlpreview.github.io/?https://github.com/upodroid/oci-proxy/blob/compliance-fixes/oci-compliance.html

/sig k8s-infra
/priority important-soon
/kind feature

**Comment by BenTheElder** [comment]
We don't support push and have no intention to, the others I don't have too much concern about, the failed responses are from AR not registry.k8s.io.

**Comment by BenTheElder** [comment]
It is worth noting that the README currently states OCI compliance, which may be misleading and also unnecessary.

Per:
https://github.com/opencontainers/distribution-spec/tree/main/conformance#testing-registry-workflows

We would need to fix the pull tests, but only the pull tests.
But again, most of these responses appear to be from AR, not archeio so .... 

Unlike Kubernetes conformance, I don't think there's a program to vet implementations and list vendors. It's unclear that these test results should be a priority vs working pull and tag list (which work fine).

**Comment by upodroid** [comment]
I did some digging around since I opened the issue:
- the conformance test runs with the v1 api which is deprecated.(There isn't a flag for v2 only tests)
- AR doesn't support the v1 API.
- AR is returning incorrect values. Blaming Google.

https://cloud.google.com/artifact-registry/docs/supported-formats

TL;DR, we meet the spec and everything is mostly good. Can't really do anything about the cloud vendor's implementation.

I don't mind putting this in a small markdown in the repo and closing out the issue.

**Comment by BenTheElder** [comment]
> the conformance test runs with the v1 api which is deprecated.(There isn't a flag for v2 only tests)
> AR doesn't support the v1 API.

Ah, yeah that will do it. We also dropped v1 from archeio for similar reasons, I don't think any clients are still on v1.

We should say that we are OCI *v2* compliant, which is also noteworthy if a v3 is released and we're not prepared yet...

For error codes, I think it's split between archeio and AR, but responses with different 4XX codes don't seem like a big deal, lots of registries are doing things like forwarding blobs to blob storage backends with their own error codes.

**Comment by BenTheElder** [comment]
Our current failures for pull-only testing (not push, delete, discovery) are the same as the AR backing registry.

They also seem like flaws in the tests, such as expecting 200 OK for some API calls even when pretty much all major registries return 3XX ...

I think we're doing a solid job implementing the intention of the API spec and we should continue to test with real API clients instead for now.
